sdk/messaging/azwebpubsub/client_custom.go

```diff
@@ -0,0 +1,170 @@
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	authPolicy := runtime.NewBearerTokenPolicy(credential, []string{internal.TokenScope}, nil)
+	azcoreClient, err := azcore.NewClient(internal.ModuleName+".Client", internal.ModuleVersion,
+		runtime.PipelineOptions{PerRetry: []policy.Policy{authPolicy}}, &options.ClientOptions)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		internal: azcoreClient,
+		endpoint: endpoint,
+		hub:      hub,
+	}, nil
+}
+
+// NewClientFromConnectionString creates a Client from a connection string.
+//
+//	Endpoint=https://<your-namespace>.webpubsub.azure.com/;AccessKey=<key>;
+func NewClientFromConnectionString(connectionString string, hub string, options *ClientOptions) (*Client, error) {
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	props, err := internal.ParseConnectionString(connectionString)
+
+	if err != nil {
+		return nil, err
+	}
+
+	azcoreClient, err := azcore.NewClient(internal.ModuleName+".Client", internal.ModuleVersion, runtime.PipelineOptions{
+		PerRetry: []policy.Policy{internal.NewWebPubSubKeyCredentialPolicy(props.AccessKey)},
+	}, &options.ClientOptions)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		internal: azcoreClient,
+		endpoint: props.Endpoint,
+		hub:      hub,
+		key:      &props.AccessKey,
+	}, nil
+}
+
+// GenerateClientAccessUrlOptions represents the options for generating a client token.
+type GenerateClientAccessUrlOptions struct {
+	// UserID is the user ID for the client.
+	UserID *string
+
+	// Roles are the roles that the connection with the generated token will have.
+	// Roles give the client initial permissions to leave, join, or publish to groups when using PubSub subprotocol.
+	// Possible role values:
+	// - webpubsub.joinLeaveGroup: the client can join or leave any group.
+	// - webpubsub.sendToGroup: the client can send messages to any group.
+	// - webpubsub.joinLeaveGroup.<group>: the client can join or leave group <group>.
+	// - webpubsub.sendToGroup.<group>: the client can send messages to group <group>.
+	// More info: https://azure.github.io/azure-webpubsub/references/pubsub-websocket-subprotocol#permissions
+	Roles []string
+
+	// ExpirationTimeInMinutes is the number of minutes until the token expires.
+	ExpirationTimeInMinutes *int32
+
+	// Groups are the groups to join when the client connects.
+	Groups []string
+}
+
+type GenerateClientAccessUrlResponse struct {
+	// The client token
+	Token string
+	// The base URL for the client to connect to
+	BaseURL string
+	// The URL client connects to with access_token query string
+	URL string
+}
+
+func (c *Client) GenerateClientAccessUrl(ctx context.Context, options *GenerateClientAccessUrlOptions) (*GenerateClientAccessUrlResponse, error) {
+	endpoint := c.endpoint
+	hubName := c.hub
+	if !strings.HasSuffix(endpoint, "/") {
+		endpoint += "/"
+	}
+	clientEndpoint := regexp.MustCompile(`$(http)(s?://)`).ReplaceAllString(endpoint, "ws$2")
```

Comment:
Do we need to define such regex as a package level var to avoid compilation every time?

Reply:
Actually I recommend `url.Parse`.